staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -671,19 +744,41 @@ func (c *Cacher) processEvent(event *watchCacheEvent) {
 }
 
 func (c *Cacher) dispatchEvents() {
+	startID := bucketID(c.clock.Now())
 	for {
 		select {
 		case event, ok := <-c.incoming:
 			if !ok {
 				return
 			}
-			c.dispatchEvent(&event)
+			if event.Type == watch.Bookmark {
+				currentID := bucketID(c.clock.Now())
+				c.dispatchBookmarkEvent(&event, startID, currentID)
+				startID = currentID + 1
+			} else {
+				c.dispatchEvent(&event)
+			}
 		case <-c.stopCh:
 			return
 		}
 	}
 }
 
+func (c *Cacher) dispatchBookmarkEvent(event *watchCacheEvent, startID, currentID int64) {
+	c.Lock()
+	defer c.Unlock()
+	for id := startID; id <= currentID; id++ {
+		watchers := c.watchers.popWatchers(id)
+		for _, w := range watchers {
+			if w.stopped {
```

Comment:
This isn't thread safe - and I'm not sure if we can even solve that safely in your current implementation in an efficient way.
The problem is that (even if you opaque this check with a lock), if stop() will be called immediately after this check but before nonBlockingAdd, stop() will close the input channel and nonBlockingAdd will panic.
And I would like to avoid blocking things for a long time.

Reply:
This problem doesn't exist today because we are careful when input channel can be closed.

I actually have a different suggestion (which is easier and less invasive into code (which makes it more safe at this point)):
Given that we only do nonblockingAdd, we can actually do this under lock. So it will look like:
- here:
`w.nonblockingAddThreadsafe(event)`
- in watcher:
```
func nonblokcingAddThreadsafe(event) {
  w.Lock()
  defer w.Unlock()
  if !w.stopped {
    switch {
    case c.input<-event
    default:
    }
}
```

Given that we close input channel under this lock, this solves the problem, and given that we don't block on putting into channel, this should be fairly fast.

Reply:
Another choice is wrap it like dispatchEvent():
startDispatching
if bookmark {
    nonblockingAdd()
} else {
    add()
}
endDispatching 

I prefer this way, it is more general.

Reply:
This highly depends how deleting a watcher would be implemented - in the current implementation (where you don't delete a watcher from those "bookmarkwatchers" together with deleting it from the rest of structure, it won't work.

Reply:
It does not matter. Wrap nonblockingAdd() with dispatching, the Stop() will be delayed at the end of the scope.

Reply:
It matters, because you choose watchers to which the event is send differently. I would need to see the whole implementation to say if it's correct or not, it may or may not be.

Reply:
`nonblokcingAddThreadsafe` certainly works. Out of curiosity, wouldn't it be better to never close the input channel and avoid the locking? Here's the idea: https://github.com/kubernetes/kubernetes/commit/e13fda875ea58e1ba7b3b34eef1bf64d338a59df. It could easily be extended for this PR.